src/Security/Security.php

```diff
@@ -303,12 +415,8 @@ public static function permissionFailure($controller = null, $messageSet = null)
                 $message = $messageSet['default'];
             }
 
-            // Somewhat hackish way to render a login form with an error message.
-            $me = new Security();
-            $form = $me->LoginForm();
-            $form->sessionMessage($message, ValidationResult::TYPE_WARNING);
-            Session::set('MemberLoginForm.force_message', 1);
-            $loginResponse = $me->login();
+            Security::singleton()->setLoginMessage($message, ValidationResult::TYPE_WARNING);
+            $loginResponse = Security::singleton()->login(new HTTPRequest('GET', '/'));
```

Comment:
I suggest to make Security::login() method take all-optional arguments, meaning $request can be null. At the moment, the $request passed in is pretty much only used to detect a custom login arg (Security/login/cms) so it should be safe enough to make it optional.

Instead of creating a new object, should we just use the singleton (with all DI configured properly?).

I.e.

$loginResponse = Security::singleton()->login();